agent/consul/state/config_entry.go

```diff
@@ -371,6 +373,100 @@ var serviceGraphKinds = []string{
 	structs.ServiceResolver,
 }
 
+// discoveryChainTargets will return a list of services listed as a target for the input's discovery chain
+func (s *Store) discoveryChainTargets(ws memdb.WatchSet, dc, service string, entMeta *structs.EnterpriseMeta) (uint64, []structs.ServiceName, error) {
+	source := structs.NewServiceName(service, entMeta)
+	req := discoverychain.CompileRequest{
+		ServiceName:         source.Name,
+		EvaluateInNamespace: source.NamespaceOrDefault(),
+
+		// TODO(freddy) : Should these be anything other than the known DC?
+		EvaluateInDatacenter: dc,
+		UseInDatacenter:      dc,
+	}
+	idx, chain, err := s.ServiceDiscoveryChain(ws, source.Name, entMeta, req)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed to fetch discovery chain for %q: %v", source.String(), err)
+	}
+
+	var resp []structs.ServiceName
+	for _, t := range chain.Targets {
+		em := structs.EnterpriseMetaInitializer(t.Namespace)
+		target := structs.NewServiceName(t.Service, &em)
+
+		// TODO (freddy): Allow upstream DC and encode in response
+		if t.Datacenter == dc {
+			resp = append(resp, target)
+		}
+	}
+	return idx, resp, nil
+}
+
+// discoveryChainSources will return a list of services whose discovery chains have the input service as a target
+func (s *Store) discoveryChainSources(ws memdb.WatchSet, tx ReadTxn, dc string, destination structs.ServiceName) (uint64, []structs.ServiceName, error) {
+	queue := []structs.ServiceName{destination}
+
+	seenLink := make(map[structs.ServiceName]bool)
```

Comment:
Should you also populate `destination` in the `seenLink` table to avoid a lookup of the destination again if it's also referenced higher up in a disco chain?

Reply:
Hmm yep, should have considered the destination to also be a valid source.